File: millancestoreapp/lib/models/order.dart
class OrderItem {
  final int id;
  final int? productId;
  final String productName;
  final String productSku;
  final int? variantId;
  final int quantity;
  final double unitPrice;
  final double totalPrice;

  const OrderItem({
    required this.id,
    this.productId,
    required this.productName,
    required this.productSku,
    this.variantId,
    required this.quantity,
    required this.unitPrice,
    required this.totalPrice,
  });

  factory OrderItem.fromJson(Map<String, dynamic> j) => OrderItem(
        id:          j['id'] as int,
        productId:   j['product_id'] as int?,
        productName: j['product_name'] as String,
        productSku:  j['product_sku'] as String,
        variantId:   j['variant_id'] as int?,
        quantity:    j['quantity'] as int,
        unitPrice:   double.parse(j['unit_price'].toString()),
        totalPrice:  double.parse(j['total_price'].toString()),
      );
}

class OrderStatusHistory {
  final String? fromStatus;
  final String toStatus;
  final String? notes;
  final DateTime changedAt;

  const OrderStatusHistory({
    this.fromStatus,
    required this.toStatus,
    this.notes,
    required this.changedAt,
  });

  factory OrderStatusHistory.fromJson(Map<String, dynamic> j) =>
      OrderStatusHistory(
        fromStatus: j['from_status'] as String?,
        toStatus:   j['to_status'] as String,
        notes:      j['notes'] as String?,
        changedAt:  DateTime.parse(j['changed_at'] as String),
      );

  String get label {
    const map = {
      'PENDING':          'Order Placed',
      'CONFIRMED':        'Order Confirmed',
      'PROCESSING':       'Processing',
      'PACKED':           'Packed',
      'SHIPPED':          'Dispatched',
      'OUT_FOR_DELIVERY': 'Out for Delivery',
      'DELIVERED':        'Delivered',
      'CANCELLED':        'Cancelled',
      'RETURN_REQUESTED': 'Return Requested',
      'RETURNED':         'Returned',
    };
    return map[toStatus] ?? toStatus;
  }
}

class Order {
  final int id;
  final String orderNumber;
  final String status;
  final double subtotal;
  final double discountAmount;
  final double taxAmount;
  final double shippingAmount;
  final double totalAmount;
  final String? shippingName;
  final String? shippingPhone;
  final String? shippingAddress;
  final String? trackingNumber;
  final String? courierName;
  final String? cancelledReason;
  final String? returnReason;
  final String? notes;
  final DateTime? confirmedAt;
  final DateTime? packedAt;
  final DateTime? dispatchedAt;
  final DateTime? deliveredAt;
  final DateTime createdAt;
  final List<OrderItem> items;
  final List<OrderStatusHistory> history;

  // Convenience getters for UI
  int get itemCount => items.length;
  bool get isActive => ![
        'DELIVERED', 'CANCELLED', 'RETURNED'
      ].contains(status);
  bool get canReturn => status == 'DELIVERED';
  bool get canCancel =>
      ['PENDING', 'CONFIRMED', 'PROCESSING'].contains(status);

  String get statusLabel {
    const map = {
      'PENDING':          'Order Placed',
      'CONFIRMED':        'Confirmed',
      'PROCESSING':       'Processing',
      'PACKED':           'Packed',
      'SHIPPED':          'Dispatched',
      'OUT_FOR_DELIVERY': 'Out for Delivery',
      'DELIVERED':        'Delivered',
      'CANCELLED':        'Cancelled',
      'RETURN_REQUESTED': 'Return Requested',
      'RETURNED':         'Returned',
    };
    return map[status] ?? status;
  }

  const Order({
    required this.id,
    required this.orderNumber,
    required this.status,
    required this.subtotal,
    required this.discountAmount,
    required this.taxAmount,
    required this.shippingAmount,
    required this.totalAmount,
    this.shippingName,
    this.shippingPhone,
    this.shippingAddress,
    this.trackingNumber,
    this.courierName,
    this.cancelledReason,
    this.returnReason,
    this.notes,
    this.confirmedAt,
    this.packedAt,
    this.dispatchedAt,
    this.deliveredAt,
    required this.createdAt,
    this.items = const [],
    this.history = const [],
  });

  factory Order.fromJson(Map<String, dynamic> j) => Order(
        id:             j['id'] as int,
        orderNumber:    j['order_number'] as String,
        status:         j['status'] as String,
        subtotal:       double.parse(j['subtotal']?.toString() ?? '0'),
        discountAmount: double.parse(j['discount_amount']?.toString() ?? '0'),
        taxAmount:      double.parse(j['tax_amount']?.toString() ?? '0'),
        shippingAmount: double.parse(j['shipping_amount']?.toString() ?? '0'),
        totalAmount:    double.parse(j['total_amount']?.toString() ?? '0'),
        shippingName:   j['shipping_name'] as String?,
        shippingPhone:  j['shipping_phone'] as String?,
        shippingAddress: j['shipping_address'] as String?,
        trackingNumber: j['tracking_number'] as String?,
        courierName:    j['courier_name'] as String?,
        cancelledReason: j['cancelled_reason'] as String?,
        returnReason:   j['return_reason'] as String?,
        notes:          j['notes'] as String?,
        confirmedAt:    j['confirmed_at'] != null ? DateTime.parse(j['confirmed_at']) : null,
        packedAt:       j['packed_at']    != null ? DateTime.parse(j['packed_at'])    : null,
        dispatchedAt:   j['dispatched_at']!= null ? DateTime.parse(j['dispatched_at']): null,
        deliveredAt:    j['delivered_at'] != null ? DateTime.parse(j['delivered_at']) : null,
        createdAt:      DateTime.parse(j['created_at'] as String),
        items: (j['items'] as List? ?? [])
            .map((e) => OrderItem.fromJson(e as Map<String, dynamic>))
            .toList(),
        history: (j['history'] as List? ?? [])
            .map((e) => OrderStatusHistory.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

File: backend/app/models/order.py
import enum
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Text, ForeignKey, Numeric
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from db.base import Base


class OrderStatus(str, enum.Enum):
    PENDING           = "PENDING"
    CONFIRMED         = "CONFIRMED"
    PROCESSING        = "PROCESSING"
    PACKED            = "PACKED"
    SHIPPED           = "SHIPPED"
    OUT_FOR_DELIVERY  = "OUT_FOR_DELIVERY"
    DELIVERED         = "DELIVERED"
    CANCELLED         = "CANCELLED"
    RETURN_REQUESTED  = "RETURN_REQUESTED"
    RETURNED          = "RETURNED"


# Valid status transitions — enforced in repository
ORDER_TRANSITIONS = {
    OrderStatus.PENDING:           [OrderStatus.CONFIRMED, OrderStatus.CANCELLED],
    OrderStatus.CONFIRMED:         [OrderStatus.PROCESSING, OrderStatus.CANCELLED],
    OrderStatus.PROCESSING:        [OrderStatus.PACKED, OrderStatus.CANCELLED],
    OrderStatus.PACKED:            [OrderStatus.SHIPPED],
    OrderStatus.SHIPPED:           [OrderStatus.OUT_FOR_DELIVERY],
    OrderStatus.OUT_FOR_DELIVERY:  [OrderStatus.DELIVERED],
    OrderStatus.DELIVERED:         [OrderStatus.RETURN_REQUESTED],
    OrderStatus.RETURN_REQUESTED:  [OrderStatus.RETURNED, OrderStatus.DELIVERED],
    OrderStatus.CANCELLED:         [],
    OrderStatus.RETURNED:          [],
}


class Order(Base):
    __tablename__ = "orders"

    id              = Column(Integer, primary_key=True, index=True, autoincrement=True)
    order_number    = Column(String(50),  unique=True, nullable=False, index=True)
    customer_id     = Column(Integer, ForeignKey("customers.id", ondelete="SET NULL"), nullable=True, index=True)
    status          = Column(String(30),  nullable=False, default=OrderStatus.PENDING)

    # Financials
    subtotal        = Column(Numeric(12, 2), nullable=False)
    discount_amount = Column(Numeric(12, 2), default=0, nullable=False)
    tax_amount      = Column(Numeric(12, 2), default=0, nullable=False)
    shipping_amount = Column(Numeric(12, 2), default=0, nullable=False)
    total_amount    = Column(Numeric(12, 2), nullable=False)
    coupon_id       = Column(Integer, ForeignKey("coupons.id", ondelete="SET NULL"), nullable=True)

    # Shipping address
    shipping_name    = Column(String(255), nullable=True)
    shipping_phone   = Column(String(20),  nullable=True)
    shipping_address = Column(Text, nullable=True)

    # Dispatch / tracking
    tracking_number  = Column(String(100), nullable=True, index=True)
    courier_name     = Column(String(100), nullable=True)

    # Stage timestamps (populated automatically on each transition)
    confirmed_at     = Column(DateTime(timezone=True), nullable=True)
    packed_at        = Column(DateTime(timezone=True), nullable=True)
    dispatched_at    = Column(DateTime(timezone=True), nullable=True)  # set when SHIPPED
    delivered_at     = Column(DateTime(timezone=True), nullable=True)

    # Misc
    notes            = Column(Text, nullable=True)
    cancelled_reason = Column(Text, nullable=True)
    return_reason    = Column(Text, nullable=True)

    created_at  = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at  = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    # Relationships
    customer = relationship("Customer", back_populates="orders")
    items    = relationship("OrderItem",         back_populates="order", cascade="all, delete-orphan")
    payment  = relationship("Payment",           back_populates="order", uselist=False)
    refunds  = relationship("Refund",            back_populates="order")
    history  = relationship("OrderStatusHistory", back_populates="order",
                            cascade="all, delete-orphan", order_by="OrderStatusHistory.changed_at")


class OrderItem(Base):
    __tablename__ = "order_items"

    id           = Column(Integer, primary_key=True, index=True, autoincrement=True)
    order_id     = Column(Integer, ForeignKey("orders.id", ondelete="CASCADE"), nullable=False, index=True)
    product_id   = Column(Integer, ForeignKey("products.id", ondelete="SET NULL"), nullable=True)
    product_name = Column(String(500), nullable=False)   # snapshot at order time
    product_sku  = Column(String(100), nullable=False)   # snapshot at order time
    variant_id   = Column(Integer, ForeignKey("product_variants.id", ondelete="SET NULL"), nullable=True)
    quantity     = Column(Integer, nullable=False)
    unit_price   = Column(Numeric(12, 2), nullable=False)
    discount_price = Column(Numeric(12, 2), nullable=True)
    tax_percent  = Column(Numeric(5, 2), default=0, nullable=False)
    total_price  = Column(Numeric(12, 2), nullable=False)

    order   = relationship("Order",   back_populates="items")
    product = relationship("Product", back_populates="order_items")


class OrderStatusHistory(Base):
    """
    Immutable log — one row per status transition.
    Lets the separate customer app (and admin) show a full timeline.
    """
    __tablename__ = "order_status_history"

    id          = Column(Integer, primary_key=True, index=True, autoincrement=True)
    order_id    = Column(Integer, ForeignKey("orders.id", ondelete="CASCADE"), nullable=False, index=True)
    from_status = Column(String(30), nullable=True)   # NULL for the first PENDING entry
    to_status   = Column(String(30), nullable=False)
    notes       = Column(Text, nullable=True)
    changed_by  = Column(Integer, ForeignKey("admins.id", ondelete="SET NULL"), nullable=True)
    changed_at  = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    order = relationship("Order", back_populates="history")
